Support Blackfly USB3 camera

andy-sweet opened the pull request:
This uses https://github.com/acquire-project/acquire-driver-egrabber as a template and modifies the driver implementation in spinnaker.cpp to support the Spinnaker SDK.

Some implementation details are specific to the Blackfly camera and will need to be generalized when targeting other cameras.

Closes #2

andy-sweet committed:
Add submodules

andy-sweet committed:
Update README and add LICENSE

andy-sweet committed:
Add gitignore

andy-sweet committed:
Add pre commit configuration

andy-sweet committed:
Add src and tests

andy-sweet commented:
Marked as draft because this probably shouldn't be merged yet, but it is generally ready to review.

I've built this and run tests locally on macOS and Windows.

The main file to focus on is `spinnaker.cpp`, as the other files were mostly unchanged from the egrabber repo. Some focus areas in the review I'd appreciate.

- Should we make the readable/writable checks more consistent for getting/setting spinnaker/GenICam fields?
- Should we always error when trying to set a non-settable field?
- Better ideas for Acquire <-> Spinnaker enum translation (compared to the switch-case converter functions I defined).
- Check out the remaining TODOs.
- Should we expect the Spinnaker SDK to be installed on the testing system? Do we need to package/distribute it here? Can we?
- Need to add GitHub actions with a runner that has a Blackfly camera connected.

andy-sweet committed:
Update core lib reference

andy-sweet committed:
Revert license copyright to template

aliddell commented on LICENSE:
```diff
@@ -186,7 +186,7 @@
       identification within third-party archives.
 
-   Copyright 2023 Chan Zuckerberg Initiative
+   Copyright [yyyy] [name of copyright owner]
```
Why change this back?

andy-sweet replied:
If you read the text above the copyright notice, I think this is intended as a template. This also explains why even Apache projects leave it like that (e.g. [kafka](https://github.com/apache/kafka/blob/trunk/LICENSE)).

Based on the same text, I think actual copyright notice referencing the owner/licensor should appear in each source code file or maybe even every file that's distributed.

Also, I'd prefer to do all the licenses at once and consistently, rather than make this a special case.

andy-sweet committed:
Add some live review comments

andy-sweet committed:
Implement exposure setter

andy-sweet committed:
Fix ID checks

andy-sweet committed:
Shutdown in constructor

andy-sweet committed:
Handle destructor exceptions

andy-sweet committed:
Genicam strings instead of spinnaker enums

andy-sweet committed:
Clean up init shutdown checks

andy-sweet committed:
Add explanatory comments

andy-sweet committed:
Remove unneeded safety in start

andy-sweet commented:
Back to draft while I address comments from the live review.

andy-sweet committed:
Reinit and clean up tests

andy-sweet committed:
Remove most readable and writable checks

andy-sweet committed:
Define const genicam strings

andy-sweet committed:
Checked set property function

andy-sweet committed:
Simplify system release

andy-sweet committed:
Configure some properties on construction

andy-sweet committed:
Reintroduce started state

andy-sweet committed:
Update comment

andy-sweet committed:
Rename property to node

andy-sweet committed:
Finish rename

andy-sweet committed:
Add TODO

andy-sweet committed:
Format code

andy-sweet committed:
Remove user from clang-tidy config

andy-sweet commented:
There are still a few TODOs and I want to add some more test coverage around setting properties, but this is generally ready for review again. I made a few commits since actually testing this with cameras, but I think those should be safe.

I am still having issues where spinnaker complains about a lack of memory when initializing the camera. I found [a way to decrease the probability of this happening](https://github.com/acquire-project/acquire-driver-spinnaker/pull/4/files#r1303325868) but it was not a complete fix.

andy-sweet commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1033 @@
+             "re-initializing.");
+        camera_->DeInit();
+    }
+    camera->Init();
```
Sometimes spinnaker can throw here, complaining that the device does not have enough memory for buffer allocation. Changing the buffer handling mode to newest only seemed to help, but did not entirely fix things.

```cpp
    Spinnaker::GenApi::INodeMap& stream_node_map = camera->GetTLStreamNodeMap();
    Spinnaker::GenApi::CEnumerationPtr stream_handling_node = stream_node_map.GetNode("StreamBufferHandlingMode");
    *stream_handling_node = "NewestOnly";
```

nclack replied:
It might be worth being opinionated about the stream handling so we get consistent behavior in the field.

andy-sweet commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1033 @@
+bool
+check_node_writable(Spinnaker::GenApi::INode& node)
+{
+    if (!Spinnaker::GenApi::IsWritable(node)) {
```
I've flip-flopped between guarding with an `if` and just `throwing`. I think the semantics of `set` mean that we want to try to set as many new parameter values as we can even if one fails, so I'm using the `if` guard right now. Maybe I should just wrap the individual setters in try/catch blocks and let spinnaker throw though?

nclack commented on src/acquire-core-libs:
looks like this might be pointing to your fork rather than [f56ba33](https://github.com/acquire-project/acquire-core-libs/commit/f56ba33ebfd5e6971ad8b7b502d76fb013de7d1e)

andy-sweet replied:
Fixed! I think... I just forgot to pull my core-lib's main and push the reference up on this branch.

andy-sweet committed:
Update core submodule

andy-sweet committed:
Fix comment

andy-sweet committed:
Update zarr submodule

andy-sweet committed:
Test software input trigger config

andy-sweet committed:
Configure exposure output in test

andy-sweet committed:
Test some properties

andy-sweet committed:
Test shape and offset

andy-sweet committed:
Reorder test to make exposure time pass

andy-sweet committed:
Fix trigger test

andy-sweet committed:
Clean up tests

andy-sweet committed:
Update submodules

andy-sweet committed:
Point submodules to main

andy-sweet committed:
Don't build submodule tests

andy-sweet committed:
Remove running check

andy-sweet committed:
Use TIFF writer instead of Zarr

andy-sweet committed:
Remove zarr submodule

andy-sweet committed:
Remove zarr from cmake

andy-sweet committed:
Clean up

andy-sweet committed:
Add GitHub CI

andy-sweet committed:
Remove linux and mac runs

andy-sweet committed:
Temp change to pull_request

andy-sweet committed:
Merge pull request #1 from andy-sweet/feat-blackfly-ci

Add GitHub CI actions

andy-sweet committed:
Format code

andy-sweet committed:
Try to trigger action

andy-sweet committed:
Revert "Try to trigger action"

This reverts commit ac6fc5b74602a2a29917e9eb7dbc082955ffb1c5.

aliddell committed:
Merge branch 'main' into feat-blackfly-usb3

aliddell committed:
Merge branch 'main' into feat-blackfly-usb3

andy-sweet commented on .github/workflows/test_pr.yml:
```diff
@@ -1,24 +1,86 @@
   push:
     branches:
       - main
   pull_request:
```
Reminder for me.

```suggestion
  pull_request_target:
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+    return memcmp(&lhs, &rhs, sizeof(Trigger)) == 0;
+}
+
+// Clamps a value to be in the closed interval [low, high].
+template<typename T>
+T
+clamp(T value, T low, T high)
+{
+    if (value < low)
+        return low;
+    if (value > high)
+        return high;
+    return value;
+}
```
Why not use `std::clamp`? Is it to avoid including `<algorithm>`?

andy-sweet replied:
No. I didn't know `clamp` was brought into `std`. Changed!

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+SpinnakerCamera::update_output_trigger_exposure(Trigger& trigger)
+{
+    trigger.kind = Signal_Output;
+    trigger.enable = (*(camera_->LineSource) == genicam_exposure_active) &&
+                     (*(camera_->LineSource) == genicam_exposure_active);
```
Why is this here twice?

andy-sweet replied:
Great question. I don't know. I swear at some point I was comparing two different things here.

Given the current hardcoding of triggers for the blackfly camera in this PR, I think this should probably be checking that the LineSelector is line 1, so I made that change. This will likely change quite a bit after adding support for the oryx camera.

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+}
+
+void
+SpinnakerCamera::query_exposure_time_capabilites(
```
```suggestion
SpinnakerCamera::query_exposure_time_capabilities(
```

andy-sweet replied:
Oops. Thanks for the catch. I fixed these locally so that both the declarations and definitions were fixed in one batch.

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+}
+
+void
+SpinnakerCamera::query_binning_capabilites(CameraPropertyMetadata* meta) const
```
```suggestion
SpinnakerCamera::query_binning_capabilities(CameraPropertyMetadata* meta) const
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+    };
+}
+void
+SpinnakerCamera::query_roi_offset_capabilites(
```
```suggestion
SpinnakerCamera::query_roi_offset_capabilities(
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+    };
+}
+void
+SpinnakerCamera::query_roi_shape_capabilites(CameraPropertyMetadata* meta) const
```
```suggestion
SpinnakerCamera::query_roi_shape_capabilities(CameraPropertyMetadata* meta) const
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+}
+
+void
+SpinnakerCamera::query_pixel_type_capabilites(
```
```suggestion
SpinnakerCamera::query_pixel_type_capabilities(
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+}
+
+void
+SpinnakerCamera::query_triggering_capabilites(CameraPropertyMetadata* meta)
```
```suggestion
SpinnakerCamera::query_triggering_capabilities(CameraPropertyMetadata* meta)
```

aliddell commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1060 @@
+    Spinnaker::CameraList camera_list = system_->GetCameras();
+    Spinnaker::CameraPtr camera =
+      camera_list.GetByIndex((unsigned int)device_id);
+    // TODO: SpinnakerCamera may partially initialize the camera
+    // and throw, so we may want a try/catch block here that deinits
+    // the camera if it was initialized.
```
When do you think you'll revisit this?

andy-sweet replied:
I'm not sure, partially because I'm not sure where this logic should live - here, in SpinnakerCamera's constructor, or in the acquire video runtime.

In general, the remaining TODOs are ones that I could not resolve on my own, so any input would be much appreciated.

Would be great to have no remaining TODOs before merging. Or if there are some, they should be tracked by GitHub issues.

aliddell replied:
This is roughly analogous to the camera restart logic we (mostly Nathan) wrote into `dcam_start`. I think it makes sense to live here, but unless you're actually seeing the problem (and / or can reproduce it in a test) I would say it's fine to leave it for now as long as you make an issue.

nclack replied:
> I'm not sure where this logic should live

The logic should live here. I suspect [`DeInit()`](http://softwareservices.flir.com/Spinnaker/latest/class_spinnaker_1_1_camera.html#ae683d1e0dd5b7166d03468c06e04907c) gets called when `Camera` destructor goes. But if you're worried about it I don't see a reason not to handle it here.

andy-sweet replied:
Moved it from here to the constructor. That didn't seem to cause any new problems, but I'll verify tomorrow.

andy-sweet committed:
Address PR feedback

andy-sweet committed:
Fix typo

andy-sweet reviewed (commented):
Thanks for the review!

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+                  .type = to_sample_type(frame->GetPixelFormatName()),
+              },
+              .hardware_timestamp = timestamp_ns,
+              .hardware_frame_id = frame_id_++,
```
better to use `frame->GetFrameID()` if you can

andy-sweet replied:
Done.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+    Spinnaker::CameraPtr camera_;
+    // Frame ID that starts at 0 each time the camera is started and
+    // increments for each frame retrieved.
+    // TODO: could use Spinnaker::Image::GetFrameID instead?
```
I commented on this below but `GetFrameID` is preferred for the hardware frame id. We want whatever is closest to the camera hardware.

andy-sweet replied:
Done.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+{
+    check_spinnaker_camera_id(i);
+
+    Spinnaker::CameraList camera_list = system_->GetCameras();
+    Spinnaker::CameraPtr camera = camera_list.GetByIndex((unsigned int)i);
+    Spinnaker::GenApi::INodeMap& nodeMap = camera->GetTLDeviceNodeMap();
+
+    const Spinnaker::GenApi::CStringPtr vendor_name =
+      nodeMap.GetNode("DeviceVendorName");
+    const Spinnaker::GenApi::CStringPtr device_name =
+      nodeMap.GetNode("DeviceModelName");
+    const Spinnaker::GenApi::CStringPtr device_sn =
+      nodeMap.GetNode("DeviceSerialNumber");
```
no change needed. But fwiw `auto` is ok.

andy-sweet replied:
With Spinnaker, I prefer being more explicit about the types because I was initially a bit confused, so going to leave this.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+    }
+
+    if (frame->IsIncomplete()) {
+        LOGE(
+          "Image incomplete: %s",
+          Spinnaker::Image::GetImageStatusDescription(frame->GetImageStatus()));
```
Might be nice to know the frame id here.

```suggestion
        LOGE(
          "Image incomplete: (%d) %s",
          (int)frame->GetFrameID(),
          Spinnaker::Image::GetImageStatusDescription(frame->GetImageStatus()));
```

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+void
+SpinnakerCamera::maybe_set_output_trigger_exposure(Trigger& target)
+{
+    // TODO: is there a way to disable an output line using Spinnaker?
```
I believe there are two ways. I like the second the most.

1. Set the LineSelector to Line0, since that line doesn't support output.
2. Set the LineSource to UserOutput1. Then based on whether you want the line to be high or low at rest, set the UserOutputValue1. It might be best to leave that undefined.

andy-sweet replied:
I like (2), so I did that. I only set UserOutputValue1 to false/off when this trigger is disabled. Don't have access to the camera right now, but I'm going to push and see what happens.

andy-sweet replied:
Not working yet. I'll debug this tomorrow.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+                     (*(camera_->LineSource) == genicam_exposure_active);
+    trigger.line = 1;
+    // TODO: check if this is the expected edge type for exposure active.
+    trigger.edge = TriggerEdge_LevelHigh;
```
I believe this is based on the `LineInverter` prop. If true, the level is low when active.

andy-sweet replied:
Done.

andy-sweet replied:
Not working yet. I'll debug this tomorrow.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+SpinnakerCamera::maybe_set_input_trigger_frame_start(Trigger& target)
+{
+    if (!is_equal(target, last_known_settings_.input_triggers.frame_start)) {
+        // Always disable trigger before any other configuration as in the
```
❤️

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+    if (target != last_known_settings_.binning) {
+        // Only one of horizontal and vertical may be writable, so explicitly
+        // check each before attempting to write.
+        // TODO: should we error/log if neither is writable?
```
I'd prefer not erroring. should log and read back what the camera is set to.

andy-sweet replied:
Done.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+{
+    if (target_us != last_known_settings_.exposure_time_us) {
+        set_float_node(camera_->ExposureTime, (double)target_us);
+        // TODO: may not need to actually get from camera because try_camera_set
```
`last_known_settings_` might be used in two ways. First, to quickly return the last known settings via get. Second, as a memoizing strategy to reduce the number of Spinnaker calls make on `set`.

I think that Second use case is the important one. And so that would be why you might want to be sure to update it here.

andy-sweet replied:
The problem with both uses is that we're assuming that Acquire is the only thing modifying the camera settings while its process is alive. That might be true in most cases, but it's not guaranteed to be true (e.g. I think I can successfully modify settings with SpinView while Acquire is running).

andy-sweet replied:
For now, I've removed the TODO and left in the read at the end of each setter.

nclack commented on src/spinnaker.cpp:
```diff
@@ -0,0 +1,1049 @@
+    const std::scoped_lock lock(lock_);
+    // Set shape before offset because Spinnaker blocks updates
+    // to offset if the shape is too big.
+    maybe_set_shape(properties->shape);
+    maybe_set_offset(properties->offset);
```
There's an order of operations thing here that I probably also have wrong in egrabber.

Depending on how the camera works you may need to set offset to (0,0) before changing the shape so you have the full range of the sensor. Then set the offset to the desired one.

andy-sweet replied:
I ran into a case like that hence the comment above. But I think you're right that I may be not handling all cases. Will ensure that tests cover those.

Do we assume that shape takes precedence over offset? E.g. what if a user has previously set a non-zero offset, then sets the shape to the same size as the sensor? In that case, I think we should refuse to set the shape and at least log an error before continuing with other properties.

nclack replied:
It's a bit like we're setting both at the same time from the perspective of the Acquire api, but spinnaker only allows changing one at a time.

The problem I've run into is when they have set a valid roi with a non-zero offset previously and then want to return to full sensor. The offset (to 0) needs to be applied before the shape change.

On the other hand, if I'm going in the other direction (full sensor to an valid offset roi), the new shape has to be applied before changing the offset.

In both cases, the user has valid settings. It's just that we need to watch the order of operations so things don't fail.

andy-sweet replied:
That's clear.

But what if they don't have valid settings? I assume we should not change either offset or shape if the combination is invalid? I.e. given Acquire treats these two as coupled the change should happen atomically or not at all (rather than with offset changed, but not shape).

andy-sweet replied:
Discussed in person and decided to only validate the region in this PR. If it's not valid, we don't try to set it.

nclack commented on cmake/git-versioning.cmake:
```diff
@@ -0,0 +1,30 @@
+endif()
+
+# Output
+add_definitions(-DGIT_TAG=${GIT_TAG})
+add_definitions(-DGIT_HASH=${GIT_HASH})
+set(CPACK_PACKAGE_VERSION ${GIT_TAG})
+message(STATUS "Version ${GIT_TAG} ${GIT_HASH}")
```
We get a lot of poop in the build logs because this is a global `add_definitions'. Maybe we can start doing better with something like this suggestion and then use the `target_add_git_versioning()` function to the main driver target. 

```suggestion
set(CPACK_PACKAGE_VERSION ${GIT_TAG})
message(STATUS "Version ${GIT_TAG} ${GIT_HASH}")
function(target_add_git_versioning tgt)
    target_compile_definitions(${tgt} 
        -DGIT_TAG=${GIT_TAG}
        -DGIT_HASH=${GIT_HASH}
    )
endfunction()
```

nclack replied:
If we figure it out here we can copy the pattern around.

andy-sweet replied:
Agreed on the poop. If that is also in egrabber (and maybe others), I'd prefer to create an issue to follow up rather than work out all the details in this PR.

andy-sweet replied:
Ah wait, now I saw the sibling comment. I effectively made those changes and the poop is gone now. How can I verify that GIT_TAG and GIT_HASH are being defined/used as intended?

andy-sweet replied:
I think I verified by grepping the build directory for those two and found that the compile commands use the correct GIT_HASH for this source, while leaving the others alone to use theirs (without the redefine). So I'm going to push this.

nclack commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+            )
+endif ()
+
+set_target_properties(${tgt} PROPERTIES
+        MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>"
+        )
```
See my comment in `git_versioning.cmake`

```suggestion
target_add_git_versioning(${tgt})
set_target_properties(${tgt} PROPERTIES
        MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>"
        )
```

nclack reviewed (changes requested):
Tried to answer questions and had minor suggestions. Marking as request changes just so we don't merge before you've had a chance to look.

andy-sweet committed:
Update src/spinnaker.cpp

Co-authored-by: Nathan Clack <nclack@gmail.com>

andy-sweet committed:
Remove frame ID member

andy-sweet committed:
Merge branch 'feat-blackfly-usb3' of github.com:andy-sweet/acquire-driver-spinnaker into feat-blackfly-usb3

andy-sweet committed:
Small fixes

andy-sweet committed:
Log when neither binning is writable

andy-sweet committed:
Use LineInverter to determine edge type

andy-sweet committed:
Define disable output exposure line

andy-sweet committed:
Try user output 0 instead of 1

andy-sweet committed:
Remove TODO after PR feedback

andy-sweet committed:
Deinit in driver open instead of constructor

andy-sweet committed:
Fix git tag/hash warnings

andy-sweet committed:
Remove fancy changes

andy-sweet committed:
Disabling output line not supported yet

andy-sweet committed:
More complex ROI logic

andy-sweet committed:
Another attempt at atomic output region

andy-sweet committed:
Give precedence to shape

andy-sweet committed:
Validate region, do not coerce

andy-sweet committed:
Read trigger edge for output.

andy-sweet committed:
Support disabling output exposure line

andy-sweet committed:
Add _ to private methods

andy-sweet commented:
@nclack : this is ready for another pass. I added a test for the ROI validation logic and implemented some logic to handle that validation. I also got some of the more advanced output line behavior implemented and tested too.

nclack reviewed (approved):
LGTM!
